src/dune_rules/gen_rules.ml

```diff
@@ -328,105 +328,120 @@ let gen_project_rules sctx project =
-            | Standalone_or_root (dir_contents, subs) ->
-              let* cctxs = gen_rules sctx dir_contents [] ~source_dir ~dir in
-              Memo.Build.parallel_iter subs ~f:(fun dc ->
-                  gen_rules sctx dir_contents cctxs ~source_dir
-                    ~dir:(Dir_contents.dir dc)
-                  >>| ignore)))
-      in
-      S.These (String.Set.of_list subdirs)
-  in
-  let subdirs_to_keep2 =
-    match components with
-    | [] ->
-      Subdir_set.These (String.Set.of_list [ ".js"; "_doc"; ".ppx"; ".dune" ])
-    | _ -> These String.Set.empty
-  in
-  Subdir_set.union_all [ subdirs_to_keep1; subdirs_to_keep2 ]
+  match components with
+  | [ ".dune"; "ccomp" ] ->
+    has_rules
+      ((* Add rules for C compiler detection *)
+       let+ () = Cxx_rules.rules ~sctx ~dir in
+       S.These String.Set.empty)
+  | ".dune" :: _ -> has_rules (Memo.Build.return (S.These String.Set.empty))
+  | ".js" :: rest ->
+    has_rules
+      (let+ () = Jsoo_rules.setup_separate_compilation_rules sctx rest in
+       match rest with
+       | [] -> S.All
+       | _ -> S.These String.Set.empty)
+  | "_doc" :: rest ->
+    has_rules
+      (let+ () = Odoc.gen_rules sctx rest ~dir in
+       match rest with
+       | [] -> S.All
+       | _ -> S.These String.Set.empty)
+  | ".ppx" :: rest ->
+    has_rules
+      (let+ () = Preprocessing.gen_rules sctx rest in
+       match rest with
+       | [] -> S.All
+       | _ -> S.These String.Set.empty)
+  | comps -> (
+    Source_tree.find_dir (Path.Build.drop_build_context_exn dir) >>= function
+    | None -> (
+      let has_rules x = has_rules (x >>> Memo.Build.return Subdir_set.empty) in
+      match List.last comps with
+      | Some ".formatted" ->
+        has_rules
+          (let* expander = Super_context.expander sctx ~dir in
+           gen_format_rules sctx ~expander ~output_dir:dir)
+      | Some ".bin" ->
+        has_rules
+          (let src_dir = Path.Build.parent_exn dir in
+           let* local_binaries =
+             Super_context.local_binaries sctx ~dir:src_dir
+           in
+           Memo.Build.sequential_iter local_binaries ~f:(fun t ->
+               let loc = File_binding.Expanded.src_loc t in
+               let src = Path.build (File_binding.Expanded.src t) in
+               let dst = File_binding.Expanded.dst_path t ~dir in
+               Super_context.add_rule sctx ~loc ~dir
+                 (Action_builder.symlink ~src ~dst)))
+      | Some ".utop" ->
+        has_rules (Utop.setup sctx ~dir:(Path.Build.parent_exn dir))
+      | Some _ -> redirect_to_parent
+      | None ->
+        (* [Source_tree.find_dir] never returns [None] for the root, so we
+           cannot be in this branch. *)
+        assert false)
+    | Some source_dir -> (
+      (* This interprets "rule" and "copy_files" stanzas. *)
+      Dir_contents.triage sctx ~dir
+      >>= function
+      | Group_part _ -> redirect_to_parent
```

Comment:
We ignore the payload of `Group_part` here and it seems we're not using anywhere else? I guess we have two options:
* Remove the payload of the `Group_part` constructor.
* Pass it on to `Redirect_to_parent` so that `load_dir` could jump directly to the right parent.
Not sure which one I prefer but I think either of them is better than the current situation.

Reply:
I wondered about the second point, but the current API seems better to me. In particular, it ensures that we can't "jump across a directory". For instance: 

- `a/b/c` redirects to `a`
- `a` generates the rules for `a/b/c`
- `a/b` has its own rules

That seems a bit weird, and likely to produce the same kind of weirdness we had before this PR.

Removing the payload of `Group_part` seems fine.